popde/adaptive_kde.py

```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+            fom += awkde.evaluate(test_data)  
+        return np.mean(fom)
+
+    def kfold_cv_score(self, bandwidth_val, alpha_val, n_splits=2):
+        """
+        to do: how to choose and use s_split in init
+        """
+        kf = KFold(n_splits=n_splits, shuffle=True, random_state=42)
+        fom = []
+        for train_index, test_index in kf.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            local_weights = None
+            awkde = AdaptiveBwKDE(train_data, local_weights, bandwidth=bandwidth_val,alpha=alpha_val)
+            fom.append(awkde.evaluate(test_data).sum())
+        return sum(fom)
+
+    def optimize_parameters(self, method='loo_cv', fom_plot=False):
+        best_score = float('inf')
+        best_params = {'bandwidth': None, 'alpha': None}
+
+        FOM= {}
+        for bandwidth in self.bandwidth_options:
+            for alpha in self.alpha_options:
+                if method=='kfold_cv':
+                    score = self.kfold_cv_score(bandwidth, alpha, n_splits=2)
+                else:
+                    score = self.loo_cv_score(bandwidth, alpha)
+
+                FOM[(bandwidth, alpha)] = score
+                if score < best_score:
+                    best_score = score
+                    best_params['bandwidth'] = bandwidth
+                    best_params['alpha'] = alpha
+        
+        if fom_plot==True:
+            import operator
+            optval = max(FOM.items(), key=operator.itemgetter(1))[0]
+            optbw, optalpha  = optval[0], optval[1]
+            maxFOM = FOM[(optbw, optalpha)]
+
+            fig = plt.figure(figsize=(12,8))
+            ax = fig.add_subplot(111)
+            for bw in self.bandwidth_options:
+                FOMlist = [FOM[(bw, al)] for al in self.alpha_options]
+                if bw not in ['silverman', 'scott']:
+                    bw = float(bw) #bwchoice.astype(np.float) #for list
+                    ax.plot(self.alpha_options, FOMlist, label='{0:.3f}'.format(bw))
+                else:
+                    ax.plot(alphagrid, FOMlist, label='{}'.format(bw))
+            if optbw not in ['silverman', 'scott']:
+                ax.plot(optalpha, maxFOM, 'ko', linewidth=10, label=r'$\alpha={0:.3f}, bw= {1:.3f}$'.format(optalpha, optbw))
+            else:
+                ax.plot(optalpha, maxFOM, 'ko', linewidth=10, label=r'$\alpha={0:.3f}, bw= {1}$'.format(optalpha, optbw))
+            ax.set_xlabel(r'$\alpha$', fontsize=18)
+            ax.set_ylabel(r'$FOM$', fontsize=18)
+            handles, labels = ax.get_legend_handles_labels()
+            lgd = ax.legend(handles, labels, loc='upper center', bbox_to_anchor=(0.5,1.25), ncol =6, fancybox=True, shadow=True, fontsize=8)
+        #plt.ylim(maxFOM -5 , maxFOM +6)
+            plt.savefig("FOMfortwoDsourcecase.png", bbox_extra_artists=(lgd, ), bbox_inches='tight')
```

Comment:
can you explain what this `bbox_extra_artists` is? 
Also, it's not a good idea to always save the plot with the same name, see my suggestion to allow different names to be chosen when calling the method.

Reply:
just to fix the legends. added a comment, a suggestion may be better